plugins/cryptolib/cryptolib.cpp

```diff
@@ -0,0 +1,705 @@
+
+    assertex(*len == currLen);
+}
+
+//-------------------------------------------------------------------------------------------------------------------------------------------
+// CIPHER SYMMETRIC SUPPORT
+//-------------------------------------------------------------------------------------------------------------------------------------------
+
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+    static CriticalSection loadingDigestCS;
+    static bool digestsLoaded = false;
+#endif
+
+void my_clAlgorithmsCallback(const OBJ_NAME * _obj, void * _pRes)
+{
+    ((StringArray*)_pRes)->append((char*)_obj->name);//add this digest to string array
+}
+
+//NB: CRYPTOLIB call signatures can be gleaned from the generated workunit CPP file
+
+//Symmetric block/stream digest algorithms
+CRYPTOLIB_API void CRYPTOLIB_CALL clInstalledSymmetricCipherAlgorithms(bool & __isAllResult, size32_t & __lenResult, void * & __result)
+{
+    __isAllResult = false;
+    StringArray algorithms;
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+    CriticalBlock block(loadingDigestCS);
+    if (!digestsLoaded)
+    {
+        OpenSSL_add_all_digests();
+        digestsLoaded = true;
+    }
+#endif
+    OBJ_NAME_do_all_sorted(OBJ_NAME_TYPE_CIPHER_METH, my_clAlgorithmsCallback, (void*)&algorithms);
+    stringArrayToECLSetOfString(algorithms, &__result, &__lenResult);
+}
+
+CRYPTOLIB_API void CRYPTOLIB_CALL clSupportedSymmetricCipherAlgorithms(bool & __isAllResult, size32_t & __lenResult, void * & __result)
+{
+    __isAllResult = false;
+    StringArray algorithms;
+    algorithms.appendList("aes-256-cbc,aes-192-cbc,aes-128-cbc", ",");//make sure this list matches loadEVP_Cipher()
+    stringArrayToECLSetOfString(algorithms, &__result, &__lenResult);
+}
+
+
+//-------------------------------------------------------------------------------------------------------------------------------------------
+
+//Build the DATA buffer returned from clSymmetricEncrypt, and sent to clSymmetricDecrypt
+//Buffer is structured as:
+// (size32_t)lenIV, IV excluding NULL, (size32_t)LenPlainText excluding NULL, (size32_t)LenCipher, Cipher
+class CSymmetricBufferHelper
+{
+public:
+    CSymmetricBufferHelper(const void * pBuffer)
+    {
+        const char * pIV;
+        size32_t lenIV;
+
+        const char * pCipher;
+        size32_t lenCipher;
+
+        char * pRes = (char *)pBuffer;
+        memcpy(&lenIV, pRes, sizeof(size32_t));
+        pRes += sizeof(size32_t);
+        pIV = pRes;
+        pRes += lenIV;
+        memcpy(&m_lenPlainText, pRes, sizeof(size32_t));
+        pRes += sizeof(size32_t);
+        memcpy(&lenCipher, pRes, sizeof(size32_t));
+        pRes += sizeof(size32_t);
+        pCipher = pRes;;
+
+        m_IV.append(lenIV, pIV);
+        m_sbCipher.append(lenCipher, pCipher);
+    }
+
+    CSymmetricBufferHelper(void * & __result, const char * pIV, size32_t lenPlainText, size32_t lenCipherBuff, const void * pCipherBuff)
```

Comment:
This would make more sense if it was serializing member variables, rather than taking parameters.  I see it is keeping the code together, but it a constructor for this class isn't the way to do it.  Possibly a deserialize function to generate a block of data from the member variables.

Reply:
I'm also hesitant that this could overwrite too much daa - since the allocation of the buffer doesn't occur in the same place, and there is no check of the size of the data that is written.